replica/distributed_loader.cc

```diff
@@ -444,11 +476,22 @@ distributed_loader::process_upload_dir(distributed<replica::database>& db, distr
             shard_gen[s].store(shard_generation_base * smp::count + s, std::memory_order_relaxed);
         }
 
+        // Pass owned_ranges_ptr to reshard to piggy-back cleanup on the resharding compaction.
+        // Note that needs_cleanup() is inaccurate and may return false positives,
+        // maybe triggerring resharding+cleanup unnecessarily for some sstables.
+        // But this is resharding on refresh (sstable loading via upload dir),
+        // which will usually require resharding anyway.
+        //
+        // FIXME: take multiple compaction groups into account
+        // - segregate resharded tables into compaction groups
+        // - split the keyspace local ranges per compaction_group as done in table::perform_cleanup_compaction
+        //   so that cleanup can be considered per compaction group
+        auto owned_ranges_ptr = compaction::make_owned_ranges_ptr(db.local().get_keyspace_local_ranges(ks));
         reshard(directory, db, ks, cf, [&global_table, upload, &shard_gen] (shard_id shard) mutable {
```

Comment:
Yes, owned_ranges_ptr is ditacting the need for cleanup. I think it will be easier to maintain if only owned_ranges_ptr drives the need for cleanup. That intention can be communicated by helper functions in compaction_descriptor.

This is similar to tombstone gc being enabled based on the presence of sstable set snapshot, see compaction_descriptor::enable_garbage_collection().

Reply:
done in v4